### PR TITLE
Add possibility to run multiple queries in one sessions, with optiona…

### PR DESCRIPTION
**Pull request by thesubneo**
…l transaction

## Summary
Introduce a possibility to run multiple queries in one session and optionally in a transaction.

## Checklist
- [x] Summarized PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-snowflake/blob/main/CHANGELOG.md)

### CODE REVIEW

File: prefect_snowflake/database.py [hunk 1]
```diff
@@ -8,6 +8,9 @@
 
 from prefect_snowflake.credentials import SnowflakeCredentials
 
+BEGIN_TRANSACTION_STATEMENT = "BEGIN"
```
Comment (desertaxle) [inline prefect_snowflake/database.py:11]: After looking at the [Snowflake docs](https://docs.snowflake.com/en/sql-reference/transactions.html#explicit-transactions), I think that this statement needs to be updated.
```suggestion
BEGIN_TRANSACTION_STATEMENT = "BEGIN TRANSACTION"
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
 ### Added
+- Support for running multiple queries in a one session and in a transaction.
```
Comment (ahuang11) [inline CHANGELOG.md:11]:
```suggestion
- Support for running multiple queries in a one session and in a transaction - [#9](https://github.com/PrefectHQ/prefect-snowflake/pull/9)
```

File: prefect_snowflake/database.py [hunk 3]
```diff
@@ -74,3 +77,89 @@ def snowflake_query_flow():
...
+    warehouse: Optional[str] = None,
+    as_transaction: bool = False,
+    return_transaction_control_results: bool = False,
+) -> List[List[Tuple[Any]]]:
+    """
+    Executes multiple queries against a Snowflake database in a shared session.
+    Allows execution in a transaction.
+
+    Args:
+        queries: The List of queries to execute against the database.
```
Comment (ahuang11) [inline prefect_snowflake/database.py:98]:
```suggestion
        queries: The list of queries to execute against the database.
```

File: prefect_snowflake/database.py [hunk 4]
```diff
@@ -74,3 +77,89 @@ def snowflake_query_flow():
...
+                user="user",
+                password="password",
+                database="database",
+                warehouse="warehouse",
+            )
+            result = snowflake_multiquery(
+                ["SELECT * FROM table WHERE id=%{id_param}s LIMIT 8;", "SELECT 1,2"],
+                snowflake_credentials,
+                params={"id_param": 1},
+                as_transaction = True
```
Comment (ahuang11) [inline prefect_snowflake/database.py:134]:
```suggestion
                as_transaction=True
```

File: prefect_snowflake/database.py [hunk 5]
```diff
@@ -74,3 +77,89 @@ def snowflake_query_flow():
...
+                while connection.is_still_running(
+                    connection.get_query_status_throw_if_error(query_id)
+                ):
+                    await asyncio.sleep(0.05)
+                cursor.get_results_from_sfqid(query_id)
+                result = cursor.fetchall()
+                results.append(result)
+
+    # cut off results from BEGIN/COMMIT queries
+    return (
```
Comment (ahuang11) [inline prefect_snowflake/database.py:161]: I think it'd be more clear if this was broken down into
```
if as_transaction and not return_transaction_control_results:
    return results[1:-1]
else:
    return results

File: prefect_snowflake/database.py [hunk 6]
```diff
@@ -74,3 +77,89 @@ def snowflake_query_flow():
...
+        queries: The List of queries to execute against the database.
+        params: The params to replace the placeholders in the query.
+        snowflake_credentials: The credentials to use to authenticate.
+        cursor_type: The type of database cursor to use for the query.
+        database: The name of the database to use; overrides
+            the credentials definition if provided.
+        warehouse: The name of the warehouse to use; overrides
+            the credentials definition if provided.
+        as_transaction: If True, queries are executed in a transaction.
+        return_transaction_control_results: Determines if the results of queries
```
Comment (ahuang11) [inline prefect_snowflake/database.py:107]: I am uncertain if this needs to be a keyword; can you provide an example when we would want the BEGIN / COMMIT statements to be returned in the result?

**Review by ahuang11 — CHANGES_REQUESTED** [review]
Thanks so much for contributing; left some comments and a question!

File: prefect_snowflake/database.py [hunk 7]
```diff
@@ -74,3 +77,89 @@ def snowflake_query_flow():
...
+            for query in queries:
+                response = cursor.execute_async(query, params=params)
+                query_id = response["queryId"]
+                while connection.is_still_running(
+                    connection.get_query_status_throw_if_error(query_id)
+                ):
+                    await asyncio.sleep(0.05)
+                cursor.get_results_from_sfqid(query_id)
+                result = cursor.fetchall()
+                results.append(result)
```
Comment (ahuang11) [inline prefect_snowflake/database.py:158]: I wonder if this should be put outside the for loop so multiple queries can run simultaneously?

**Comment by ahuang11** [comment]
Some things to note:
https://docs.snowflake.com/en/user-guide/python-connector-example.html#best-practices-for-asynchronous-queries

File: prefect_snowflake/database.py (hunk 7 above)
Comment (ahuang11) [reply prefect_snowflake/database.py:158]: Maybe this task should have a `run_sequentially=True/False` keyword
https://docs.snowflake.com/en/user-guide/python-connector-example.html#best-practices-for-asynchronous-queries

Comment (thesubneo) [reply prefect_snowflake/database.py:158]: I'm not sure if it provides any additional functionality. Running these transaction simultaneously can be easily achieved in user code by running `snowflake_multiquery ` task simultaneously...

File: prefect_snowflake/database.py (hunk 6 above)
Comment (thesubneo) [reply prefect_snowflake/database.py:107]: They return `Statement executed successfully.` so I imagine that there is a remote possibility that they return something else. Then user code could react to this, but I don't know real life scenario. I also use it in unit tests. Shall If remove this?

File: prefect_snowflake/database.py (hunk 5 above)
Comment (thesubneo) [reply prefect_snowflake/database.py:161]: ok

File: prefect_snowflake/database.py (hunk 6 above)
Comment (desertaxle) [reply prefect_snowflake/database.py:107]: I think the default to not return the transaction control results is great and the ability to return them if needed is reasonable, so I say we keep it.

File: prefect_snowflake/database.py (hunk 7 above)
Comment (desertaxle) [reply prefect_snowflake/database.py:158]: I think we should assume that the statements executed in this task are dependent and recommend users to run multiple tasks if the statements are not dependent like @thesubneo suggested.

**Review by desertaxle — APPROVED** [review]

**Review by ahuang11 — APPROVED** [review]

**Pull request merged**